src/main/java/bio/terra/app/controller/SnapshotsApiController.java

```diff
@@ -171,27 +174,30 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
 
   @Override
   public ResponseEntity<EnumerateSnapshotModel> enumerateSnapshots(
-      @Valid @RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
-      @Valid @RequestParam(value = "limit", required = false, defaultValue = "10") Integer limit,
-      @Valid @RequestParam(value = "sort", required = false, defaultValue = "created_date")
-          EnumerateSortByParam sort,
-      @Valid @RequestParam(value = "direction", required = false, defaultValue = "asc")
-          SqlSortDirection direction,
-      @Valid @RequestParam(value = "filter", required = false) String filter,
-      @Valid @RequestParam(value = "region", required = false) String region,
-      @Valid @RequestParam(value = "datasetIds", required = false) List<String> datasetIds) {
+      Integer offset,
+      Integer limit,
+      EnumerateSortByParam sort,
+      SqlSortDirection direction,
+      String filter,
+      String region,
+      List<String> datasetIds) {
     ControllerUtils.validateEnumerateParams(offset, limit);
-    Set<UUID> resources =
-        iamService
-            .listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT)
-            .keySet();
+    Map<UUID, Set<IamRole>> idsAndRoles =
+        iamService.listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT);
     List<UUID> datasetUUIDs =
         ListUtils.emptyIfNull(datasetIds).stream()
             .map(UUID::fromString)
             .collect(Collectors.toList());
     EnumerateSnapshotModel edm =
         snapshotService.enumerateSnapshots(
-            offset, limit, sort, direction, filter, region, datasetUUIDs, resources);
+            offset, limit, sort, direction, filter, region, datasetUUIDs, idsAndRoles.keySet());
+    for (SnapshotSummaryModel summary : edm.getItems()) {
```

Comment:
This is a lot of logic to live in the controller.  I think it makes sense to push the logic down into the the service layer

Reply:
I can do that, but it means that the service layer will now need to know about SAM, which it didn't need to know about before.

Reply:
Yeah, though I don't think that's a huge deal.  We already import the IamService into a few service classes

Reply:
And now it's no longer just doing gatekeeping but actually augmenting the objects